Add a fchownat(2) wrapper

jmmv opened the pull request:
The existing chown(2) wrapper should be removed in favor of this, but I do
not yet know if fchmodat(2) is available in all the platforms we may care
about.  Will consider later.

jmmv commented:
Per the TODO, do you think `chown` should be removed?

The way I'd validate that is: add a trivial test for `fchownat` to ensure we reference this function somewhere and then ask CI to validate all platforms. If that succeeds, I can later try to see if this is sufficient for older macOS versions, which still exist in Travis, to prevent the issue we saw with `utimes`.

I can do this before the merge; just let me know if there are other considerations.

asomers commented on src/fcntl.rs:
```diff
@@ -182,6 +183,15 @@ pub fn readlinkat<'a, P: ?Sized + NixPath>(dirfd: RawFd, path: &P, buffer: &'a m
 
+/// Computes the raw fd consumed by a function of the form `*at`.
+#[inline]
+pub fn at_rawfd(fd: Option<RawFd>) -> raw::c_int {
```
Does this need to be public?  Will Nix consumers ever need to use it directly?

asomers replied:
If you need to use it from within multiple of Nix's modules, then maybe you should be using `pub(crate)` or `pub(super)`.

jmmv replied:
Oh, I wasn't aware of such a thing. Done.

asomers commented on src/unistd.rs:
```diff
@@ -557,27 +557,89 @@ pub fn getcwd() -> Result<PathBuf> {
 
+/// Computes the raw UID and GID values to pass to a `*chown` call.
+///
+/// According to the POSIX specification, -1 is used to indicate that owner and group
```
This part doesn't belong in the doc comment.  You can make it a regular comment instead.

asomers commented on src/unistd.rs:
```diff
@@ -557,27 +557,89 @@ pub fn getcwd() -> Result<PathBuf> {
+/// According to the POSIX specification, -1 is used to indicate that owner and group
+/// are not to be changed.  Since uid_t and gid_t are unsigned types, we have to wrap
+/// around to get -1.
+#[inline]
```
`#[inline]` doesn't do anything for non-public functions, so you can remove it.

asomers commented on src/unistd.rs:
```diff
@@ -557,27 +557,89 @@ pub fn getcwd() -> Result<PathBuf> {
+/// # References
+///
+/// [fchownat(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fchownat.html).
+#[inline]
```
I favor not using `#[inline]` at all.  Application authors can already use LTO if they really care.

jmmv commented:
I don't mind keeping `chown`, but then would the same rationale apply for `lchown` and we should add it? Why doesn't the same apply for other traditional functions like `utimes`?

asomers commented:
> I don't mind keeping `chown`, but then would the same rationale apply for `lchown` and we should add it? Why doesn't the same apply for other traditional functions like `utimes`?
> 
> In fact, I think that if the OS exposes such functions, they should be exposed in `nix` as well. `nix` shouldn't be in the business of calling "traditional but ugly" functions obsolete... because all the new-ish counterparts are not always available :(

The same rationale would apply to `lchown` only if we already provide it.  I don't want to withdraw `chown` because consumers may already be using it.  `lchown`, OTOH, is unnecessary unless somebody ports Nix to a weird platform that doesn't have `fchownat`.

jmmv commented:
I've added tests to ensure these functions exists in all CI platforms, and also confirmed in my project that `fchownat` exists for the older macOS that I need. So no need to add `lchown` from my point of view, which is what I wanted to confirm.

asomers commented on src/fcntl.rs:
```diff
@@ -182,6 +183,15 @@ pub fn readlinkat<'a, P: ?Sized + NixPath>(dirfd: RawFd, path: &P, buffer: &'a m
 }
 
+/// Computes the raw fd consumed by a function of the form `*at`.
+#[inline]
```
Again, no need for `#[inline]` for private functions.

asomers commented on test/test_unistd.rs:
```diff
@@ -301,6 +301,45 @@ fn test_getcwd() {
+
+    let tempdir = tempfile::tempdir().unwrap();
+    let path = tempdir.path().join("file");
+    drop(File::create(&path).unwrap());
```
As in the `ftruncate` review, I prefer using lexical scope to `drop`.

jmmv commented:
Updated the commit message and removed the original contents of the PR message, which I think would be added to the history as well.

Also rebased on top of the truncate merge to avoid `CHANGELOG` conflicts.
